Realm/RLMArray.mm

```diff
@@ -32,19 +32,39 @@ @implementation RLMArray {
     NSMutableArray *_backingArray;
 }
 
-- (instancetype)initWithObjectClassName:(NSString *)objectClassName standalone:(BOOL)standalone {
+static void changeArray(__unsafe_unretained RLMArray *const ar, NSKeyValueChange kind, NSIndexSet *is, dispatch_block_t f) {
+    if (!ar->_backingArray) {
+        ar->_backingArray = [NSMutableArray new];
+    }
+
+    if (RLMObjectBase *parent = ar->_parentObject) {
+        [parent willChange:kind valuesAtIndexes:is forKey:ar->_key];
+        f();
+        [parent didChange:kind valuesAtIndexes:is forKey:ar->_key];
+    }
+    else {
+        f();
+    }
+}
+
+static void changeArray(__unsafe_unretained RLMArray *const ar, NSKeyValueChange kind, NSUInteger index, dispatch_block_t f) {
+    changeArray(ar, kind, [NSIndexSet indexSetWithIndex:index], f);
```

Comment:
When I first wrote it willChange/didChange were always called so there was no point, and I forgot to update that.